core/src/main/scala/sttp/tapir/Validator.scala

```diff
@@ -160,75 +200,130 @@ object Validator extends ValidatorMacros {
   }
+  case class Any[T](validators: immutable.Seq[Validator[T]]) extends Validator[T] {
 
-  //
-
-  def show[T](v: Validator[T]): Option[String] = {
-    v match {
-      case Min(value, exclusive) => Some(s"${if (exclusive) ">" else ">="}$value")
-      case Max(value, exclusive) => Some(s"${if (exclusive) "<" else "<="}$value")
-      // TODO: convert to patterns when https://github.com/lampepfl/dotty/issues/12226 is fixed
-      case p: Pattern[T]                     => Some(s"~${p.value}")
-      case m: MinLength[T]                   => Some(s"length>=${m.value}")
-      case m: MaxLength[T]                   => Some(s"length<=${m.value}")
-      case m: MinSize[T, _]                  => Some(s"size>=${m.value}")
-      case m: MaxSize[T, _]                  => Some(s"size<=${m.value}")
-      case Custom(_, showMessage)            => showMessage.orElse(Some("custom"))
-      case Enumeration(possibleValues, _, _) => Some(s"in(${possibleValues.mkString(",")}")
-      case Mapped(wrapped, _)                => show(wrapped)
-      case All(validators) =>
-        validators.flatMap(show(_)) match {
-          case immutable.Seq()  => None
-          case immutable.Seq(s) => Some(s)
-          case ss               => Some(s"all(${ss.mkString(",")})")
-        }
-      case Any(validators) =>
-        validators.flatMap(show(_)) match {
-          case immutable.Seq()  => Some("reject")
-          case immutable.Seq(s) => Some(s)
-          case ss               => Some(s"any(${ss.mkString(",")})")
-        }
+    override def show: Option[String] = validators.flatMap(_.show) match {
+      case Nil      => Some("reject")
+      case x :: Nil => Some(x)
+      case xs       => Some(s"any(${xs.mkString(",")})")
     }
+
+    override def apply(t: T): ValidationResult[T] =
+      ValidationResult.partitionMap(
+        validators.toList.map(_.apply(t))
+      ) match {
+        case (_, _ :: _)    => ValidationResult.Valid(t)
+        case (invalid, Nil) => ValidationResult.Invalid(t, invalid.flatMap(_.errors))
+      }
+
+    override def contramap[TT](g: TT => T): Validator[TT] =
+      if (validators.isEmpty) Any(Nil) else super.contramap(g)
+
+    override def or(other: Validator[T]): Validator[T] =
+      if (validators.isEmpty) other else Any(validators :+ other)
   }
 }
 
-sealed trait ValidationError[T] {
-  def prependPath(f: FieldName): ValidationError[T]
-  def invalidValue: T
-  def path: List[FieldName]
+sealed trait ValidationResult[+T] {
+  def valid[TT >: T]: Option[Valid[TT]] =
+    this match {
+      case v: Valid[TT]   => Some(v)
+      case _: Invalid[TT] => None
+    }
+
+  def invalid[TT >: T]: Option[Invalid[TT]] =
+    this match {
+      case _: Valid[TT]   => None
+      case i: Invalid[TT] => Some(i)
+    }
 }
+object ValidationResult {
+  case class Valid[T](value: T) extends ValidationResult[T]
+  case class Invalid[T](value: T, errors: List[ValidationError]) extends ValidationResult[T] {
```

Comment:
I think I'd prefer to retain the validator that caused the error as part of the `ValidationError` - currently in the new design we are loosing this information, in favour of constructing the error description earlier. I'd defer creating the string to a later time, and

Reply:
This was one of my biggest doubts about this refactoring. 
In the end, I've used this design for those reasons: 
- The Validator info is not used actually (ideally it's great to have it but actually is useless and introduce to many compromises)
- The early creation of the error message allows more flexibility to custom validator 

Adding the `Validator` info to `ValidationError ` is trivial but introduce some problems for the custom validators, check this example:

Reply:
Ok thanks, let me try to unpack this so that we're on the same page :)

Not sure what you mean by validator info not being used - it was being used in `ValdiatorMessages` to create the appropriate message?

I do get the point with the custom validator - so far the message structure was fixed (unless you overwrote the `DecodeFailureHandler`) to `s"expected $valueName to pass custom validation: ${c.message}, but was '${ve.invalidValue}'"`, and you'd like to customise this?

Maybe we could have both with a structure like this:

```scala
case class ValidationError[T](v: Validator[T], invalidValue: T, path: FieldPath, customMessage: Option[String])`
```

Then, we retain all of the information, and given a `ValdiationError` we can create a nice error message - using the default structure, unless there's a custom message defined, which entirely replaces the default one (which is a change from what we have now). What do you think?

Reply:
> Not sure what you mean by validator info not being used - it was being used in ValdiatorMessages to create the appropriate message?

Yes exactly, but just to do pattern matching and create the right error message. But If now we create the error message when we are defining the validator this is no needed

---

> Maybe we could have both with a structure like this

The problem is that you don't have a `Vallidator` instance when you are defining it, this problem exists only for custom validators

Given this
```scala
case class Custom[T](doValidate: T => ValidationResult[T], showMessage: Option[String] = None) extends Validator[T]
case class ValidationError[T](v: Validator[T], invalidValue: T, path: FieldPath, customMessage: Option[String])
```

We ha some problems doing this, moreover we are duplicating the `invalidValue`(both inside `Invalid` and `ValidationError`)
```scala
val myCustomVal : Validator[Int] = Custom[Int](myNumb match {
   case 1 => ValidationResult.Valid(myNumb)
   case _ => ValidationResult.Invalid(myNumb, List(
      ValidationError(???, myNumb, ???, Some("Number must be 1"))
  ))
}
```

Reply:
Another proposal, instead of delegating the creation of the full error message to each `Validator` why we don't just use a `String` for describing the cause of the error and create the full error message just changing a little bit the template

Like this
```scala
case class ValidationError(cause: String, path: FieldNames = FieldNames.empty) {

  def prependPath(f: FieldName): ValidationError =
    copy(path = path.prependPath(f))

  def message: String = s"Validation failed at `${path.asPath.getOrElse(FieldPath("value"))}` because: $cause"
}
```


So the previous example becomes
```scala
val myCustomVal : Validator[Int] = Custom[Int](myNumb match {
   case 1 => ValidationResult.Valid(myNumb)
   case _ => ValidationResult.Invalid(myNumb, List(
      ValidationError(s"Number must be 1 but was $myNumb")
  ))
}

myCustomVal
   .validate(0)
   .invalid
   .get
   .description // Validation failed for `value` because: Number must be 1 but was 0
```

We can even remove `ValidationMessage` in this way

Reply:
You're right, good point with the self-reference when creating a custom validator. I remember now having a similar problem originally ;)

Sorry for my stubbornness though, but I will insist that we keep the creation of the messages remains as part of the `DecodeFailureHandler`. It's a feature of tapir - that users can customise the errors - that I want to retain.

Maybe at the expense of making creation of custom validators slightly less convenient, we could pass in functions which would mark the value as valid / invalid, sth like:

```scala
case class Custom[T](doValidate: ValidationContext[T] => ValidationResult[T]) extends Validator[T]

trait ValidationContext[T] {
  def value: T
  def valid: ValidationResult[T]
  def invalid(errorMessage: Option[String]): ValidationResult[T]
}

case class ValidationError(path: FieldPath)
// ValidationResult.Invalid would contain an optional errorMessages
```

these functions would then be able to capture the self-reference to the `Validator` instance and use them in the result.

Reply:
> Sorry for my stubbornness though, but I will insist that we keep the creation of the messages remains as part of the DecodeFailureHandler. It's a feature of tapir - that users can customise the errors - that I want to retain.

You are right! I've changed my mind about this. In fact, in my last code snippet the final message is inside tapir and not customizable

This actually can be moved inside `DecodeFailureHandler`
```scala
def message: String = s"Validation failed at `${path.asPath.getOrElse(FieldPath("value"))}` because: $cause"
```

Have a look at `Circe`, this is the way how they define errors 
```scala
sealed abstract class DecodingFailure(val reason: Reason) extends Error {
   def history: List[CursorOp]
    
   val message: String = reason match {
   case WrongTypeExpectation(expJsType, v) => s"Got value '${v.noSpaces}' with wrong type, expecting $expJsType"
   case MissingField                       => "Missing required field"
   case CustomReason(message)              => message
  }
```


----
`ValidationContext ` is a good solution if we want to keep the `Validator` instance 👍. But IMHO once we wrap the error template, there a no reason to keep it because the error message(or better the cause not the full error message) is created in the definition of the validator. What do think? 

If it can helps I can update the PR with this solution so maybe is more clear my point 😊

Reply:
Hm how can we have customisable error messages (preferrably as part of `DecodeFailureHandler`), without having access to the `Validator` instance?

Reply:
We just need the `ValidationResult` and not the `Validator` instance to build the error message. If the results contain the reason(as Circe do) we can build the error message in `DecodeFailureHandler` just using the result of the validation 

for example:
```scala
case class ValidationError(expectedValue: T, reason: String)

//DecodeFailureHandler
val validationResult: ValidationResult[T] = ???
val errors: Option[List[String]] = validationResult.invalid.map(invalid => invalid.errors.map(e => {
  s"Validation failed at `${e.path.asPath.getOrElse(FieldPath("value"))}` because: ${e.cause}, expected ${e.expectedValue}"
  })
)
```

Reply:
Well but you can't e.g. customise the message for `Min` specifically. Let's say I'd like to change `expected $valueName to be greater than` to polish or some other exotic language ;)

In the above code, only the general template which is used to present validation errors is customisable. I think that anything that ends up being sent to the user as a string should be customisable, or at leat we should try :)

Reply:
Ok i got you! and what do you think I we have the possibility to recover the validator ? In this way we can re-map the error message 

```scala 
Min(1, false).recoverWith(_.withError(ValidationError("Must be non empty!")))
```

Reply:
Hm I suppose to answer that we have to consider where you'd prefer to customize the error messages: centrally in `DecodeFailureHandler`, or individually per-validator usage.

I guess there are use-cases for both, and as I think we already noted in the PR, we do need a way to customise the error message entirely. But (excuse me for my stubbornness) I think that central configuration is more often useful - so that you don't have to always remember to use a customised-error-message-`Min`-validator, you just customize it once when the message is generated.

Reply:
Ok make sense, let's use 'Validator' instance 👍